### PR TITLE
block modification when one node building

### PR DESCRIPTION
**Pull request by etiennehomer**
Signed-off-by: Etienne Homer <etiennehomer@gmail.com>

### CODE REVIEW

**Review by laurgarn — APPROVED** [review]
Nothing suspicious found

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 1]
```diff
@@ -2372,6 +2383,14 @@ public void testCreateSubstation() throws Exception {
                 studyNameUserIdUuid, modificationNode1Uuid, MODIFICATION_UUID).content(substationAttributesUpdated).contentType(MediaType.APPLICATION_JSON))
             .andExpect(status().isOk());
 
+        String createSubstationAttributes2 = "{\"substationId\":\"substationId2\",\"substationName\":\"substationName2\",\"country\":\"AD\"}";
+        modificationNode1.setBuildStatus(BuildStatus.BUILDING);
+        networkModificationTreeService.updateNode(studyNameUserIdUuid, modificationNode1);
+        // create substation on building node
+        mockMvc.perform(post("/v1/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/substations",
+                        studyNameUserIdUuid, modificationNode1Uuid).content(createSubstationAttributes2))
+                .andExpect(status().is4xxClientError());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:2392]: isForbidden() ?

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 2]
```diff
@@ -2423,6 +2442,15 @@ public void testCreateVoltageLevel() throws Exception {
...
             .andExpect(status().isOk());
 
+        String createVoltageLevelAttributes2 = "{\"voltageLevelId\":\"voltageLevelId3\",\"voltageLevelName\":\"voltageLevelName3\""
+                + ",\"nominalVoltage\":\"379.1\", \"substationId\":\"s2\"}";
+        modificationNode1.setBuildStatus(BuildStatus.BUILDING);
+        networkModificationTreeService.updateNode(studyNameUserIdUuid, modificationNode1);
+        // create voltage level on building node
+        mockMvc.perform(post("/v1/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/voltage-levels",
+                        studyNameUserIdUuid, modificationNode1Uuid).content(createVoltageLevelAttributes2))
+                .andExpect(status().is4xxClientError());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:2452]: isForbidden() ?

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 3]
```diff
@@ -2752,6 +2780,14 @@ public void testCreateGenerator() throws Exception {
                 studyNameUserIdUuid, modificationNode1Uuid, MODIFICATION_UUID).content(generatorAttributesUpdated))
             .andExpect(status().isOk());
 
+        String createGeneratorAttributes2 = "{\"generatorId\":\"generatorId3\",\"generatorName\":\"generatorName3\",\"energySource\":\"UNDEFINED\",\"minActivePower\":\"100.0\",\"maxActivePower\":\"200.0\",\"ratedNominalPower\":\"50.0\",\"activePowerSetpoint\":\"10.0\",\"reactivePowerSetpoint\":\"20.0\",\"voltageRegulatorOn\":\"true\",\"voltageSetpoint\":\"225.0\",\"voltageLevelId\":\"idVL1\",\"busOrBusbarSectionId\":\"idBus1\"}";
+        modificationNode1.setBuildStatus(BuildStatus.BUILDING);
+        networkModificationTreeService.updateNode(studyNameUserIdUuid, modificationNode1);
+        // create generator on building node
+        mockMvc.perform(post("/v1/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/generators",
+                        studyNameUserIdUuid, modificationNode1Uuid).content(createGeneratorAttributes2))
+                .andExpect(status().is4xxClientError());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:2789]: isForbidden() ?

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 4]
```diff
@@ -2787,6 +2823,14 @@ public void testCreateShuntsCompensator() throws Exception {
                 studyNameUserIdUuid, modificationNode1Uuid, MODIFICATION_UUID).content(shuntCompensatorAttributesUpdated))
             .andExpect(status().isOk());
 
+        String createShuntCompensatorAttributes2 = "{\"shuntCompensatorId\":\"shuntCompensatorId3\",\"shuntCompensatorName\":\"shuntCompensatorName3\",\"voltageLevelId\":\"idVL1\",\"busOrBusbarSectionId\":\"idBus1\"}";
+        modificationNode1.setBuildStatus(BuildStatus.BUILDING);
+        networkModificationTreeService.updateNode(studyNameUserIdUuid, modificationNode1);
+        // create shunt compensator on building node
+        mockMvc.perform(post("/v1/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/shunt-compensators",
+                        studyNameUserIdUuid, modificationNode1Uuid).content(createShuntCompensatorAttributes2))
+                .andExpect(status().is4xxClientError());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:2832]: isForbidden() ?

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 5]
```diff
@@ -2845,6 +2889,19 @@ public void testCreateLine() throws Exception {
...
+                + "\"shuntConductance1\":\"100.0\"," + "\"shuntSusceptance1\":\"100.0\","
+                + "\"shuntConductance2\":\"200.0\"," + "\"shuntSusceptance2\":\"200.0\","
+                + "\"voltageLevelId1\":\"idVL1\"," + "\"busOrBusbarSectionId1\":\"idBus1\","
+                + "\"voltageLevelId2\":\"idVL2\"," + "\"busOrBusbarSectionId2\":\"idBus2\"}";
+        modificationNode1.setBuildStatus(BuildStatus.BUILDING);
+        networkModificationTreeService.updateNode(studyNameUserIdUuid, modificationNode1);
+        // create line on building node
+        mockMvc.perform(post("/v1/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/line",
+                        studyNameUserIdUuid, modificationNode1Uuid).content(createLineAttributes2))
+                .andExpect(status().is4xxClientError());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:2903]: isForbidden() ?

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 6]
```diff
@@ -2889,6 +2946,14 @@ public void testCreateTwoWindingsTransformer() throws Exception {
                 studyNameUserIdUuid, modificationNode1Uuid, MODIFICATION_UUID).content(twoWindingsTransformerAttributesUpdated))
             .andExpect(status().isOk());
 
+        String createTwoWindingsTransformerAttributes2 = "{\"equipmentId\":\"2wtId3\",\"equipmentName\":\"2wtName3\",\"seriesResistance\":\"10\",\"seriesReactance\":\"10\",\"magnetizingConductance\":\"100\",\"magnetizingSusceptance\":\"100\",\"ratedVoltage1\":\"480\",\"ratedVoltage2\":\"380\",\"voltageLevelId1\":\"CHOO5P6\",\"busOrBusbarSectionId1\":\"CHOO5P6_1\",\"voltageLevelId2\":\"CHOO5P6\",\"busOrBusbarSectionId2\":\"CHOO5P6_1\"}";
+        modificationNode1.setBuildStatus(BuildStatus.BUILDING);
+        networkModificationTreeService.updateNode(studyNameUserIdUuid, modificationNode1);
+        // create Two Windings Transformer on building node
+        mockMvc.perform(post("/v1/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/two-windings-transformers",
+                        studyNameUserIdUuid, modificationNode1Uuid).content(createTwoWindingsTransformerAttributes2))
+                .andExpect(status().is4xxClientError());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:2955]: isForbidden() ?

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 7]
```diff
@@ -2237,6 +2240,14 @@ public void testCreateLoad() throws Exception {
                 studyNameUserIdUuid, modificationNode1Uuid, MODIFICATION_UUID).content(loadAttributesUpdated).contentType(MediaType.APPLICATION_JSON))
             .andExpect(status().isOk());
 
+        String createLoadAttributes2 = "{\"loadId\":\"loadId3\",\"loadName\":\"loadName3\",\"loadType\":\"UNDEFINED\",\"activePower\":\"100.0\",\"reactivePower\":\"50.0\",\"voltageLevelId\":\"idVL1\",\"busId\":\"idBus1\"}";
+        modificationNode3.setBuildStatus(BuildStatus.BUILDING);
+        networkModificationTreeService.updateNode(studyNameUserIdUuid, modificationNode3);
+        // create load on building node
+        mockMvc.perform(post("/v1/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/loads",
+                        studyNameUserIdUuid, modificationNode3Uuid).content(createLoadAttributes2))
+                .andExpect(status().is4xxClientError());
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/test/java/org/gridsuite/study/server/StudyTest.java]: Done

File: src/test/java/org/gridsuite/study/server/StudyTest.java (hunk 1 above)
Comment (etiennehomer) [reply src/test/java/org/gridsuite/study/server/StudyTest.java]: Done

File: src/test/java/org/gridsuite/study/server/StudyTest.java (hunk 2 above)
Comment (etiennehomer) [reply src/test/java/org/gridsuite/study/server/StudyTest.java]: Done

File: src/test/java/org/gridsuite/study/server/StudyTest.java (hunk 3 above)
Comment (etiennehomer) [reply src/test/java/org/gridsuite/study/server/StudyTest.java]: Done

File: src/test/java/org/gridsuite/study/server/StudyTest.java (hunk 4 above)
Comment (etiennehomer) [reply src/test/java/org/gridsuite/study/server/StudyTest.java]: Done

File: src/test/java/org/gridsuite/study/server/StudyTest.java (hunk 5 above)
Comment (etiennehomer) [reply src/test/java/org/gridsuite/study/server/StudyTest.java]: Done

File: src/test/java/org/gridsuite/study/server/StudyTest.java (hunk 6 above)
Comment (etiennehomer) [reply src/test/java/org/gridsuite/study/server/StudyTest.java]: Done

**Review by FranckLecuyer — APPROVED** [review]

**Pull request merged**